### PR TITLE
2270 support new rpc endpoint ticket balance and all ticket balance

### LINKED ISSUE #2270: [Mumbai] Support new RPC endpoints in Mumbai update
Mumbai protocol update introduces these 2 new RPC additions:

- Add RPC to get contract’s balance of ticket with specified ticketer, content type, and content. Can be used for both implicit and originated contracts. **POST** 
`/chains/<chain_id>/blocks/<block_id>/context/contracts/<contract_id>/ticket_balance`
MR [!6488](https://gitlab.com/tezos/tezos/-/merge_requests/6488)

- Add RPC to get the complete list of tickets owned by a given contract by scanning the contract’s storage. Can only be used for originated contracts. **POST** 
`/chains/<chain_id>/blocks/<block_id>/context/contracts/<contract_id>/all_ticket_balances`
MR [!6712](https://gitlab.com/tezos/tezos/-/merge_requests/6712)

### Solution
Implement these 2 new RPC endpoints in the `taquito/rpc` package and add relevant unit and integration tests
- for integration tests, also will need to originate the contract that does hold tickets

relevant links:
- https://tezos.gitlab.io/protocols/016_mumbai.html#rpc-changes

### PR DESCRIPTION
**Pull request by hui-an-yang**
closes #2270

- support new rpc endpoint ticket_balance and all_ticket_balances in @taquito/rpc package

Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [x] Your code builds cleanly without any errors or warnings
- [x] You have run the linter against the changes
- [x] You have added unit tests (if relevant/appropriate)
- [x] You have added integration tests (if relevant/appropriate)
- [x] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [x] You have added or updated corresponding documentation
- [x] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

In this PR, please also make sure: 

- [x] You have linked this PR to the issue by putting `closes #TICKETNUMBER` in the description box (when applicable)
- [x] You have added a concise description on your changes
## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

### CODE REVIEW

File: integration-tests/originate-known-contracts.ts [hunk 1]
```diff
@@ -135,7 +135,7 @@ CONFIGS().forEach(({ lib, setup, protocol }) => {
     });
 
     // originate tx rollup
-    if (protocol === Protocols.ProtoALpha) {
+    if (protocol === Protocols.ProtoALpha || protocol === Protocols.PtMumbaii) {
```
Comment (dsawali) [inline integration-tests/originate-known-contracts.ts:138]: I think we might not need this anymore since TORU is now disabled in Mumbai(?)

File: packages/taquito-rpc/src/rpc-client-interface.ts [hunk 2]
```diff
@@ -119,6 +121,12 @@ export interface RpcClientInterface {
   ): Promise<TxRollupInboxResponse | null>;
   getStorageUsedSpace(contract: string, options?: RPCOptions): Promise<string>;
   getStoragePaidSpace(contract: string, options?: RPCOptions): Promise<string>;
+  getTicketBalance(
+    contract: string,
+    ticket: ticketTokenParams,
```
Comment (dsawali) [inline packages/taquito-rpc/src/rpc-client-interface.ts:126]: `ticketTokenParams` should be uppercase since it's a type

File: packages/taquito-rpc/src/rpc-client-interface.ts [hunk 3]
```diff
@@ -119,6 +121,12 @@ export interface RpcClientInterface {
...
+    options?: RPCOptions
+  ): Promise<string>;
+  getAllTicketBalances(contract: string, options?: RPCOptions): Promise<allTicketBalancesResponse>;
```
Comment (dsawali) [inline packages/taquito-rpc/src/rpc-client-interface.ts:129]: `allTicketBalancesResponse` should be uppercase `AllTicketBalancesResponse`

File: packages/taquito-rpc/src/types.ts [hunk 4]
```diff
@@ -1148,6 +1148,20 @@ export interface PackDataParams {
   gas?: BigNumber;
 }
 
+export interface ticketTokenParams {
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1151]: same comment as above

File: packages/taquito-rpc/src/types.ts [hunk 5]
```diff
@@ -1148,6 +1148,20 @@ export interface PackDataParams {
...
+  ticketer: string;
+  content_type: MichelsonV1Expression;
+  content: MichelsonV1Expression;
+}
+
+export interface allTicketBalancesResponse {
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1157]: same comment as above

File: packages/taquito-rpc/src/types.ts [hunk 6]
```diff
@@ -1148,6 +1148,20 @@ export interface PackDataParams {
...
+  ticketer: string;
+  content_type: MichelsonV1Expression;
+  content: MichelsonV1Expression;
+  amount: string;
+}
+[];
```
Comment (dsawali) [inline packages/taquito-rpc/src/types.ts:1163]: I think instead we should make a single `TicketBalance` type, and have `AllTicketBalancesResponse` be an array of that type

**Review by dsawali — CHANGES_REQUESTED** [review]

File: integration-tests/originate-known-contracts.ts (hunk 1 above)
Comment (hui-an-yang) [reply integration-tests/originate-known-contracts.ts:138]: If we run this file now it will try to originate tx rollup contract in mumbainet is the reason I added that condition, to just leave it as empty string like ProtoAlpha is now. It will be even better we remove tx rollup origination script from that file and remove the references in config of integration-tests and all integration tests.

Comment (dsawali) [reply integration-tests/originate-known-contracts.ts:138]: yeah exactly, I think we should perhaps remove it if the contract isn't even being used. Might need to double check for that though

Comment (hui-an-yang) [reply integration-tests/originate-known-contracts.ts:138]: I didn't remove the txrollup script there cuz I wasn't sure how much it will affect our integration-test config and related integration tests. Is this something we'd like to create a ticket for or I just ship it with this pr?

File: integration-tests/rpc-nodes.spec.ts [hunk 7]
```diff
@@ -433,6 +444,18 @@ CONFIGS().forEach(
...
         });
+
+        MumbaiAndAlpha('Verify that rpcClient.ticketBalance will retrieve the specified ticket owned by the given contract', async (done) => {
+          const ticketBalance = await rpcClient.getTicketBalance(ticketContract.address, { ticketer: ticketContract.address, content_type: { prim: 'string' }, content: { string: 'abc' } } );
+          expect(ticketBalance).toBeDefined();
+          done();
+        });
+  
+        MumbaiAndAlpha('Verify that rpcClient.allTicketBalances will retrieve all tickets owned by the given contract', async (done) => {
+          const ticketBalances = await rpcClient.getAllTicketBalances(ticketContract.address!);
```
Comment (zainen) [inline integration-tests/rpc-nodes.spec.ts:455]: this doesnt give a typescript error? i thought we didnt allow non-null assertions

Comment (hui-an-yang) [reply integration-tests/rpc-nodes.spec.ts]: Don't know why it didn't complaint, but good catch thank you!! It's removed now

File: integration-tests/data/contract_origination.ts [hunk 8]
```diff
@@ -1,69 +1,5 @@
 import { OperationContentsOrigination } from "@taquito/rpc"
 
-export const contractOriginationParams = {
```
Comment (dsawali) [inline integration-tests/data/contract_origination.ts:3]: was this related to toru? wondering why it was removed

File: integration-tests/rpc-nodes.spec.ts [hunk 9]
```diff
@@ -16,16 +17,33 @@ CONFIGS().forEach(
     knownBigMapContract,
     knownSaplingContract,
     knownViewContract,
-    txRollupAddress,
   }) => {
     const Tezos = lib;
 
     const Limanet = protocol === Protocols.PtLimaPtL ? it : it.skip;
-
+    const MumbaiAndAlpha = protocol === Protocols.PtMumbaii || protocol === Protocols.ProtoALpha ? test : test.skip;
```
Comment (dsawali) [inline integration-tests/rpc-nodes.spec.ts:24]: nitpick, but M should be lowercase since it's a function and not a class/type.

File: packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts [hunk 10]
```diff
@@ -292,7 +293,8 @@ export class RpcClientCache implements RpcClientInterface {
     const key = this.formatCacheKey(
       this.rpcClient.getRpcUrl(),
       RPCMethodName.GET_NORMALIZED_SCRIPT,
-      [block, address, unparsingMode]
+      [block, address],
+      unparsingMode
```
Comment (dsawali) [inline packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts:297]: why was this moved out of the array?

File: packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts [hunk 11]
```diff
@@ -936,10 +939,12 @@ export class RpcClientCache implements RpcClientInterface {
...
   ): Promise<{ packed: string; gas: BigNumber | 'unaccounted' | undefined }> {
-    const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), RPCMethodName.PACK_DATA, [
-      block,
-      data,
-    ]);
+    const key = this.formatCacheKey(
+      this.rpcClient.getRpcUrl(),
+      RPCMethodName.PACK_DATA,
+      [block],
+      data
```
Comment (dsawali) [inline packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts:946]: same comment as above

File: packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts [hunk 12]
```diff
@@ -1165,4 +1170,58 @@ export class RpcClientCache implements RpcClientInterface {
       return response;
     }
   }
+
+  /**
+   *
+   * @param contract address of the contract we want to retrieve ticket balance of
+   * @param ticket object to specify ticketer, content type and content
```
Comment (dsawali) [inline packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts:1177]: I think we can specify it to be 'Ticket token parameter object that contains ticketer, content type, and content'

File: packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts [hunk 13]
```diff
@@ -1165,4 +1170,58 @@ export class RpcClientCache implements RpcClientInterface {
...
+    } else {
+      const response = this.rpcClient.getTicketBalance(contract, ticket, { block });
+      this.put(key, response);
+      return response;
+    }
+  }
+
+  /**
+   *
+   * @param contract address of the contract we want to retrieve ticket balance of
```
Comment (dsawali) [inline packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts:1205]: it's probably a better description to say: '...retrieve all ticket balances from'

File: packages/taquito-rpc/test/taquito-rpc.spec.ts [hunk 14]
```diff
@@ -50,6 +50,7 @@ import {
   delegatesKathmandunetSample,
   votingInfoKathmandunetSample,
   ticketUpdatesSample,
+  ticketBalances,
```
Comment (dsawali) [inline packages/taquito-rpc/test/taquito-rpc.spec.ts:53]: might want to rename this to ticketBalancesResponse to make it a bit more descriptive

**Review by dsawali — CHANGES_REQUESTED** [review]

**Comment by dsawali** [comment]
I'm not sure if integration tests were added, but I couldn't find any

File: packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts (hunk 11 above)
Comment (hui-an-yang) [reply packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts:946]: cuz data is a request body not a param this is the formatCacheKey signature ```private formatCacheKey(
    rpcUrl: string,
    rpcMethodName: RPCMethodName,
    rpcMethodParams: RpcMethodParam[],
    rpcMethodData?: object
  )```

File: packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts (hunk 10 above)
Comment (hui-an-yang) [reply packages/taquito-rpc/src/rpc-client-modules/rpc-cache.ts:297]: cuz data is a request body not a param this is the formatCacheKey signature ```private formatCacheKey(
    rpcUrl: string,
    rpcMethodName: RPCMethodName,
    rpcMethodParams: RpcMethodParam[],
    rpcMethodData?: object
  )```

File: integration-tests/data/contract_origination.ts (hunk 8 above)
Comment (hui-an-yang) [reply integration-tests/data/contract_origination.ts:3]: line12 tx_rollup_l2_address don't think it's possible to originate this contract since it's disabled

File: integration-tests/rpc-nodes.spec.ts (hunk 9 above)
Comment (hui-an-yang) [reply integration-tests/rpc-nodes.spec.ts:24]: also this file line 443 and onward is the integration test :)

File: integration-tests/rpc-nodes.spec.ts [hunk 15]
```diff
@@ -433,6 +439,18 @@ CONFIGS().forEach(
           expect(paidSpace).toBeDefined();
           done();
         });
+
+        MumbaiAndAlpha('Verify that rpcClient.ticketBalance will retrieve the specified ticket owned by the given contract', async (done) => {
+          const ticketBalance = await rpcClient.getTicketBalance(ticketContract.address, { ticketer: ticketContract.address, content_type: { prim: 'string' }, content: { string: 'abc' } } );
+          expect(ticketBalance).toBeDefined();
```
[the remark this answers is not in the document]
Comment (hui-an-yang) [reply integration-tests/rpc-nodes.spec.ts:445]: since ticket balance will just be a string, think i'll leave this one as is.

File: integration-tests/rpc-nodes.spec.ts [hunk 16]
```diff
@@ -433,6 +439,18 @@ CONFIGS().forEach(
...
+          done();
+        });
+  
+        MumbaiAndAlpha('Verify that rpcClient.allTicketBalances will retrieve all tickets owned by the given contract', async (done) => {
+          const ticketBalances = await rpcClient.getAllTicketBalances(ticketContract.address);
+          expect(ticketBalances).toBeDefined();
```
[the remark this answers is not in the document]
Comment (hui-an-yang) [reply integration-tests/rpc-nodes.spec.ts:451]: good catch, will add them accordingly

**Review by dsawali — APPROVED** [review]
